File: src/providers/chatgpt/conversations.ts
/**
 * ChatGPT conversations extraction + transform strategies.
 */
import { Readable } from "node:stream";
import { parser } from "stream-json";
import { streamArray } from "stream-json/streamers/StreamArray";

import type { ExtractionStrategy, TransformStrategy } from "../../core/etl.js";
import type { TaskMetadata, ThreadRow } from "../../core/types.js";
import type { StorageBackend } from "../../storage/backend.js";
import { ChatGPTMessageSchema } from "./schemas.js";
import {
  CURRENT_THREAD_PAYLOAD_VERSION,
  makeApplication,
  makeCollection,
  makeTextMessage,
  makeSendMessage,
  makeReceiveMessage,
  uniqueKeySuffix,
  toDict,
  getSendMessagePreview,
  getReceiveMessagePreview,
  type FibreSendMessage,
  type FibreReceiveMessage,
  type CollectionType,
} from "../../payload/models.js";

const CHATGPT_APPLICATION = makeApplication("assistant");
const CHUNK_SIZE = 500;

/** Timestamps above this threshold are treated as milliseconds (year 2100+). */
const MAX_SECONDS_EPOCH = 4_102_444_800;

function safeTimestamp(ts: number | null | undefined): Date | null {
  if (ts == null) return null;
  let val = Number(ts);
  if (val > MAX_SECONDS_EPOCH) val /= 1000;
  return new Date(val * 1000);
}

// ---------------------------------------------------------------------------
// Extraction
// ---------------------------------------------------------------------------

export class ChatGPTConversationsExtractionStrategy
  implements ExtractionStrategy
{
  async extract(
    task: TaskMetadata,
    storage: StorageBackend,
  ): Promise<Record<string, any>[][]> {
    const key = task.filenames[0];
    const data = await storage.read(key);

    return new Promise<Record<string, any>[][]>((resolve, reject) => {
      const batches: Record<string, any>[][] = [];
      let chunk: Record<string, any>[] = [];
      let rejected = false;
      const onError = (err: Error) => {
        if (!rejected) {
          rejected = true;
          reject(err);
        }
      };

      const nodeStream = Readable.from(Buffer.from(data));
      const jsonParser = parser();
      const arrayStream = streamArray();

      nodeStream.on("error", onError);
      jsonParser.on("error", onError);
      arrayStream.on("error", onError);

      const pipeline = nodeStream.pipe(jsonParser).pipe(arrayStream);

      pipeline.on("data", ({ value: conversation }: { value: any }) => {
        const conversationTitle = conversation.title ?? null;
        const conversationId = conversation.conversation_id ?? null;
        const mapping: Record<string, any> = conversation.mapping ?? {};

        for (const mappingItem of Object.values(mapping)) {
          const messageData = (mappingItem as any).message;
          if (!messageData) continue;
          if (!messageData.author || !messageData.content) continue;

          const content = messageData.content ?? {};
          if (content.content_type !== "text") continue;

          const parsed = ChatGPTMessageSchema.safeParse(messageData);
          if (!parsed.success) continue;
          const msg = parsed.data;

          if (msg.author.role === "system") continue;
          if (!msg.content.parts || !msg.content.parts[0]) continue;
          const text = msg.content.parts[0];
          if (!text.trim()) continue;

          const createTime = msg.create_time ?? null;

          chunk.push({
            role: msg.author.role,
            content: text,
            create_time: createTime,
            conversation_id: conversationId,
            conversation_title: conversationTitle,
            source: JSON.stringify(messageData),
          });

          if (chunk.length >= CHUNK_SIZE) {
            batches.push(chunk);
            chunk = [];
          }
        }
      });

      pipeline.on("end", () => {
        if (chunk.length > 0) batches.push(chunk);
        resolve(batches);
      });

      pipeline.on("error", onError);
    });
  }
}

// ---------------------------------------------------------------------------
// Transform
// ---------------------------------------------------------------------------

export class ChatGPTConversationsTransformStrategy
  implements TransformStrategy
{
  async transform(
    task: TaskMetadata,
    batches: Record<string, any>[][],
  ): Promise<ThreadRow[][]> {
    const resultBatches: ThreadRow[][] = [];

    for (const batch of batches) {
      const rows: ThreadRow[] = [];
      for (const record of batch) {
        const payload = this.buildPayload(record);
        if (!payload) continue;

        const createTime = record.create_time;
        const asat = safeTimestamp(createTime) ?? new Date();

        const dict = toDict(payload);
        rows.push({
          uniqueKey: `chatgpt_conversations:${uniqueKeySuffix(dict)}`,
          provider: task.provider,
          interactionType: task.interactionType,
          preview:
            payload.fibre_kind === "SendMessage"
              ? getSendMessagePreview(payload as FibreSendMessage, "ChatGPT")
              : getReceiveMessagePreview(
                  payload as FibreReceiveMessage,
                  "ChatGPT",
                ),
          payload: dict,
          source: record.source ?? null,
          version: CURRENT_THREAD_PAYLOAD_VERSION,
          asat,
          assetUri: null,
        });
      }
      if (rows.length > 0) resultBatches.push(rows);
    }

    return resultBatches;
  }

  private buildPayload(
    record: Record<string, any>,
  ): FibreSendMessage | FibreReceiveMessage | null {
    const role: string = record.role;
    const content: string = record.content;
    const conversationTitle: string | null = record.conversation_title ?? null;
    const conversationId: string | null = record.conversation_id ?? null;

    let context: CollectionType | undefined;
    if (conversationTitle || conversationId) {
      context = makeCollection({
        name: conversationTitle ?? undefined,
        id: conversationId
          ? `https://chatgpt.com/c/${conversationId}`
          : undefined,
      });
    }

    const message = makeTextMessage(content, context);

    const createTime = record.create_time;
    const published = safeTimestamp(createTime);
    const publishedStr = published?.toISOString();

    if (role === "user") {
      return makeSendMessage({
        object: message,
        target: CHATGPT_APPLICATION,
        published: publishedStr,
      });
    } else if (role === "assistant") {
      return makeReceiveMessage({
        object: message,
        actor: CHATGPT_APPLICATION,
        published: publishedStr,
      });
    }

    return null;
  }
}

File: context_use/providers/chatgpt/conversations.py
from __future__ import annotations

import json
import logging
from collections.abc import Iterator
from datetime import UTC, datetime
from decimal import Decimal

import ijson

from context_use.batch.grouper import CollectionGrouper
from context_use.etl.core.pipe import Pipe
from context_use.etl.core.types import ThreadRow
from context_use.etl.models.etl_task import EtlTask
from context_use.etl.payload.models import (
    CURRENT_THREAD_PAYLOAD_VERSION,
    Application,
    Collection,
    FibreReceiveMessage,
    FibreSendMessage,
    FibreTextMessage,
)
from context_use.memories.config import MemoryConfig
from context_use.memories.prompt.conversation import ConversationMemoryPromptBuilder
from context_use.providers.chatgpt.schemas import (
    ChatGPTConversationRecord,
    ChatGPTMessage,
)
from context_use.providers.types import InteractionConfig
from context_use.storage.base import StorageBackend

logger = logging.getLogger(__name__)

CHATGPT_APPLICATION = Application(name="assistant")  # type: ignore[reportCallIssue]

# Timestamps above this threshold are treated as milliseconds (year 2100+)
_MAX_SECONDS_EPOCH = 4_102_444_800  # 2100-01-01 00:00 UTC


def _safe_timestamp(ts: float | int | None) -> datetime | None:
    """Convert a Unix epoch to datetime, handling ms-vs-s ambiguity."""
    if ts is None:
        return None
    ts = float(ts)
    if ts > _MAX_SECONDS_EPOCH:
        ts /= 1000.0
    return datetime.fromtimestamp(ts, tz=UTC)


def _build_payload(
    record: ChatGPTConversationRecord,
) -> FibreSendMessage | FibreReceiveMessage | None:
    """Build an ActivityStreams payload from a conversation record."""
    # Build conversation context
    context = None
    if record.conversation_title or record.conversation_id:
        ctx_kwargs: dict = {"type": "Collection"}
        if record.conversation_title:
            ctx_kwargs["name"] = record.conversation_title
        if record.conversation_id:
            ctx_kwargs["id"] = f"https://chatgpt.com/c/{record.conversation_id}"
        context = Collection(**ctx_kwargs)

    message = FibreTextMessage(content=record.content, context=context)  # type: ignore[reportCallIssue]

    published = _safe_timestamp(record.create_time)

    if record.role == "user":
        return FibreSendMessage(  # type: ignore[reportCallIssue]
            object=message,
            target=CHATGPT_APPLICATION,
            published=published,
        )
    elif record.role == "assistant":
        return FibreReceiveMessage(  # type: ignore[reportCallIssue]
            object=message,
            actor=CHATGPT_APPLICATION,
            published=published,
        )

    return None


class ChatGPTConversationsPipe(Pipe[ChatGPTConversationRecord]):
    """ETL pipe for ChatGPT conversations.

    Reads ``conversations.json`` via ijson streaming, yields individual
    :class:`ChatGPTConversationRecord` instances, and transforms each
    into a :class:`ThreadRow` with an ActivityStreams payload.
    """

    provider = "chatgpt"
    interaction_type = "chatgpt_conversations"
    archive_version = "v1"
    archive_path_pattern = "conversations.json"
    record_schema = ChatGPTConversationRecord

    def extract(
        self,
        task: EtlTask,
        storage: StorageBackend,
    ) -> Iterator[ChatGPTConversationRecord]:
        stream = storage.open_stream(task.source_uri)

        try:
            for conversation in ijson.items(stream, "item"):
                conversation_title = conversation.get("title")
                conversation_id = conversation.get("conversation_id")
                mapping = conversation.get("mapping", {})

                for _msg_id, mapping_item in mapping.items():
                    message_data = mapping_item.get("message")
                    if not message_data:
                        continue
                    if "author" not in message_data or "content" not in message_data:
                        continue

                    content = message_data.get("content", {})
                    if content.get("content_type") != "text":
                        continue

                    try:
                        parsed = ChatGPTMessage.model_validate(message_data)
                    except Exception:
                        continue

                    # Skip roles we can't map to a payload
                    if parsed.author.role not in ("user", "assistant"):
                        continue
                    if not parsed.content.parts or not parsed.content.parts[0]:
                        continue
                    text = parsed.content.parts[0]
                    if not text.strip():
                        continue

                    # ijson may return Decimal – coerce for JSON serialization
                    create_time = parsed.create_time
                    if isinstance(create_time, Decimal):
                        create_time = float(create_time)

                    yield ChatGPTConversationRecord(
                        role=parsed.author.role,
                        content=text,
                        create_time=create_time,
                        conversation_id=conversation_id,
                        conversation_title=conversation_title,
                        source=json.dumps(message_data, default=str),
                    )
        finally:
            stream.close()

    def transform(
        self,
        record: ChatGPTConversationRecord,
        task: EtlTask,
    ) -> ThreadRow:
        payload = _build_payload(record)
        assert payload is not None, (
            f"Unexpected None payload for role={record.role!r}; "
            "extract() should have filtered this record"
        )

        asat = _safe_timestamp(record.create_time) or datetime.now(UTC)

        return ThreadRow(
            unique_key=f"chatgpt_conversations:{payload.unique_key_suffix()}",
            provider=self.provider,
            interaction_type=self.interaction_type,
            preview=payload.get_preview("ChatGPT") or "",
            payload=payload.to_dict(),
            version=CURRENT_THREAD_PAYLOAD_VERSION,
            asat=asat,
            source=record.source,
        )


INTERACTION_CONFIG = InteractionConfig(
    pipe=ChatGPTConversationsPipe,
    memory=MemoryConfig(
        prompt_builder=ConversationMemoryPromptBuilder,
        grouper=CollectionGrouper,
    ),
)
